### PR TITLE
Added api unit tests. Decoupled base_endpoint.

### PR DESCRIPTION
**Pull request by maximrodin**
Added basic rest_qa_api tests

Changed response validation order - headers first
Fixed error with nested steps in error message
Fixed error when json list does not have value (added Index Error handler)
Fixed error with list usage as json payload
Added logic to forbid Models usage without @scaf_dataclass decorator

### CODE REVIEW

File: common/libs/exception_thread/ex_thread.py [hunk 1]
```diff
@@ -2,10 +2,10 @@
 import traceback
 import queue
 from multiprocessing import TimeoutError
-import logging
```
Comment (temichus) [inline common/libs/exception_thread/ex_thread.py:5]: As I told. I don't like this solution. wat about replacing for lobs  -logger = logging.getLogger("thread") to common.scaf.get_logger(__name__) under the hood?

File: common/rest_qa_api/base_endpoint.py [hunk 2]
```diff
@@ -650,7 +706,7 @@ def __init__(self, base_url):
         self.response_model = response_model
         self.make_url_method = make_url_method
         # Dummy container to prepare request fields
-        self._request = type("Jon_Golt", (object,), dict())()
+        self._request = type("Jon_Galt", (object,), dict())()
```
Comment (temichus) [inline common/rest_qa_api/base_endpoint.py:709]: may be type("Dummy_Container", (object,), dict())() ?

Comment (maximrodin) [reply common/rest_qa_api/base_endpoint.py:709]: nope, it's a reference ;)

**Comment by maximrodin** [comment]
Hi @temichus, unfortunately, we don't have unit tests coverage yet cause it's just  a very beginning of unit tests implementation and I will add it later during integration with CI per commit. Thx for your review. About builtin logging per lib with additional logger override in Facade - it sound good and I will make this change in scope of another PR.

File: README.md [hunk 3]
```diff
@@ -43,6 +43,12 @@ sections = api web
 [api]
 api_url = http://api.openweathermap.org/data/2.5/
 
+[api_validation]
+validate_status_code = True
+validate_headers = True
+validate_body = True
+fail_if_field_is_missing = True
```
Comment (codci) [inline README.md:50]: probably rename to: validate_is_field_missing (to have the same naming for all checks)

File: common/libs/exception_thread/ex_thread.py (hunk 1 above)
Comment (maximrodin) [reply common/libs/exception_thread/ex_thread.py:5]: will be fixed in next PR

File: README.md (hunk 3 above)
Comment (maximrodin) [reply README.md]: Fixed

File: common/libs/config_parser/section/api_validation_section.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            self.config.add_section(self.SECTION_NAME)
+        super()._load_from_config()
+
+    def to_dict(self):
+        """Convert to dictionary."""
+        return {field: getattr(self, field, None) for field in self._settings}
+
+    def _perform_custom_tunings(self):
+        """Perform custom tunings for obtained settings."""
+        for setting in self._settings:
```
[the remark this answers is not in the document]
Comment (maximrodin) [reply common/libs/config_parser/section/api_validation_section.py:50]: Will fix in the scope of the other PR and general for all sections

**Comment by maximrodin** [comment]
Closed because of mixed commits from diff branch. Clean PR created with fixed comments https://github.com/Softeq/SCAF/pull/8

**Pull request closed without merging**